ledger/eval.go

```diff
@@ -1218,39 +1221,188 @@ func eval(ctx context.Context, l ledgerForEvaluator, blk bookkeeping.Block, vali
+// The order of the transaction groups returned by the channel is identical to the one in the input array.
+func loadAccounts(ctx context.Context, l ledgerForEvaluator, rnd basics.Round, groups [][]transactions.SignedTxnWithAD, feeSinkAddr basics.Address) chan loadedTransactionGroup {
+	outChan := make(chan loadedTransactionGroup, len(groups))
+	go func() {
+		// groupTask helps to organize the account loading for each transaction group.
+		type groupTask struct {
+			// balances contains the loaded balances each transaction group have
+			balances []basics.BalanceRecord
+			// balancesCount is the number of balances that nees to be loaded per transaction group
+			balancesCount int
+			// done is a waiting channel for all the account data for the transaction group to be loaded
+			done chan error
 		}
-		state.lookup(addr)
-	}
-	for _, stxn := range payset {
-		select {
-		case <-ctx.Done():
-			return
-		default:
+		// addrTask manage the loading of a single account address.
+		type addrTask struct {
+			// account address to fetch
+			address basics.Address
+			// a list of transaction group tasks that depends on this address
+			groups []*groupTask
+			// a list of indices into the groupTask.balances where the address would be stored
+			groupIndices []int
 		}
-		state.lookup(stxn.Txn.Sender)
-		maybelookup(stxn.Txn.Receiver)
-		maybelookup(stxn.Txn.CloseRemainderTo)
-		maybelookup(stxn.Txn.AssetSender)
-		maybelookup(stxn.Txn.AssetReceiver)
-		maybelookup(stxn.Txn.AssetCloseTo)
-		maybelookup(stxn.Txn.FreezeAccount)
-		for _, xa := range stxn.Txn.Accounts {
-			maybelookup(xa)
+		defer close(outChan)
+
+		accountsChannels := make(map[basics.Address]*addrTask)
+		addressesCh := make(chan *addrTask, len(groups)*16*10)
+		// totalBalances counts the total number of balances over all the transaction groups
+		totalBalances := 0
+		initAccount := func(addr basics.Address, wg *groupTask) {
+			if addr.IsZero() {
+				return
+			}
+			if task, have := accountsChannels[addr]; !have {
+				task := &addrTask{
+					address:      addr,
+					groups:       make([]*groupTask, 1, 4),
+					groupIndices: make([]int, 1, 4),
+				}
+				task.groups[0] = wg
+				task.groupIndices[0] = wg.balancesCount
+				accountsChannels[addr] = task
+				addressesCh <- task
+			} else {
+				task.groups = append(task.groups, wg)
+				task.groupIndices = append(task.groupIndices, wg.balancesCount)
+			}
+			wg.balancesCount++
+			totalBalances++
 		}
-	}
+		// add the fee sink address to the accountsChannels/addressesCh so that it will be loaded first.
+		if len(groups) > 0 {
+			task := &addrTask{
+				address: feeSinkAddr,
+			}
+			addressesCh <- task
+			accountsChannels[feeSinkAddr] = task
+		}
+
+		// iterate over the transaction groups and add all their account addresses to the list
+		groupsReady := make([]*groupTask, len(groups))
+		for i, group := range groups {
+			groupWg := &groupTask{}
```

Comment:
what does `groupWg` mean? **wg** might mean **work group** but then the abbreviation is similar to **wait group** from `sync` package.

Reply:
I changes it to be just "task"